app/src/main/java/com/rootstrap/android/ui/activity/main/SignInActivity.kt

```diff
@@ -47,46 +52,54 @@ class SignInActivity : PermissionActivity(), AuthView {
     }
 
     private fun signIn() {
-        val user = User(
-            email = email_edit_text.value(),
-            password = password_edit_text.value()
-        )
-        viewModel.signIn(user)
+        val name = binding.emailEditText.value()
+        val pass = binding.passwordEditText.value()
+
+        if (viewModel.canSignIn(name, pass)) {
+            val user = User(
+                email = name,
+                password = pass
+            )
+            viewModel.signIn(user)
+        } else {
+            showLoginError()
+        }
+    }
+
+    private fun showLoginError() {
+        binding.passwordTextInputLayout.error = getString(R.string.login_failed)
+        binding.emailTextInputLayout.error = " "
     }
 
     // ViewModelListener
     private val viewModelListener = object : ViewModelListener {
         override fun updateState() {
             when (viewModel.state) {
-                SignInState.signInFailure -> showError(viewModel.error)
+                SignInState.signInFailure, SignInState.none -> showLoginError()
                 SignInState.signInSuccess -> showProfile()
-                else -> {
-                }
             }
         }
 
         override fun updateNetworkState() {
             when (viewModel.networkState) {
                 NetworkState.loading -> showProgress()
                 NetworkState.idle -> hideProgress()
-                else -> showError(viewModel.error ?: getString(R.string.default_error))
+                else -> {
+                    hideProgress()
+                    if (viewModel.error.isNullOrEmpty())
+                        showError(getString(R.string.default_error))
+                }
             }
         }
     }
 
-    fun sampleAskForPermission() {
+    private fun sampleAskForPermission() {
         requestPermission(arrayOf(Manifest.permission.CAMERA), object : PermissionResponse {
-            override fun granted() {
-                // TODO..
-            }
+            override fun granted() = Unit
 
-            override fun denied() {
-                // TODO..
-            }
+            override fun denied() = Unit
 
-            override fun foreverDenied() {
-                // TODO..
-            }
+            override fun foreverDenied() = Unit
         })
     }
```

Comment:
this is an example function about how to ask permissions using PermissionActivity, removed it if not needed.

Reply:
done!